crates/ev-revm/src/handler.rs

```diff
@@ -0,0 +1,162 @@
+        self.redirect
+    }
+}
+
+impl<EVM, ERROR, FRAME> Handler for EvHandler<EVM, ERROR, FRAME>
+where
+    EVM: EvmTr<Context: ContextTr<Journal: JournalTr<State = EvmState>>, Frame = FRAME>,
+    ERROR: EvmTrError<EVM>,
+    FRAME: FrameTr<FrameResult = FrameResult, FrameInit = FrameInit>,
+{
+    type Evm = EVM;
+    type Error = ERROR;
+    type HaltReason = HaltReason;
+
+    fn validate_env(&self, evm: &mut Self::Evm) -> Result<(), Self::Error> {
+        self.inner.validate_env(evm)
+    }
+
+    fn validate_initial_tx_gas(&self, evm: &Self::Evm) -> Result<InitialAndFloorGas, Self::Error> {
+        self.inner.validate_initial_tx_gas(evm)
+    }
+
+    fn load_accounts(&self, evm: &mut Self::Evm) -> Result<(), Self::Error> {
+        self.inner.load_accounts(evm)
+    }
+
+    fn apply_eip7702_auth_list(&self, evm: &mut Self::Evm) -> Result<u64, Self::Error> {
+        self.inner.apply_eip7702_auth_list(evm)
+    }
+
+    fn validate_against_state_and_deduct_caller(
+        &self,
+        evm: &mut Self::Evm,
+    ) -> Result<(), Self::Error> {
+        self.inner.validate_against_state_and_deduct_caller(evm)
+    }
+
+    fn first_frame_input(
+        &mut self,
+        evm: &mut Self::Evm,
+        gas_limit: u64,
+    ) -> Result<FRAME::FrameInit, Self::Error> {
+        self.inner.first_frame_input(evm, gas_limit)
+    }
+
+    fn last_frame_result(
+        &mut self,
+        evm: &mut Self::Evm,
+        frame_result: &mut <FRAME as FrameTr>::FrameResult,
+    ) -> Result<(), Self::Error> {
+        self.inner.last_frame_result(evm, frame_result)
+    }
+
+    fn run_exec_loop(
+        &mut self,
+        evm: &mut Self::Evm,
+        first_frame_input: <FRAME as FrameTr>::FrameInit,
+    ) -> Result<FrameResult, Self::Error> {
+        self.inner.run_exec_loop(evm, first_frame_input)
+    }
+
+    fn eip7623_check_gas_floor(
+        &self,
+        evm: &mut Self::Evm,
+        exec_result: &mut <FRAME as FrameTr>::FrameResult,
+        init_and_floor_gas: InitialAndFloorGas,
+    ) {
+        self.inner
+            .eip7623_check_gas_floor(evm, exec_result, init_and_floor_gas)
+    }
+
+    fn refund(
+        &self,
+        evm: &mut Self::Evm,
+        exec_result: &mut <FRAME as FrameTr>::FrameResult,
+        eip7702_refund: i64,
+    ) {
+        self.inner.refund(evm, exec_result, eip7702_refund)
+    }
+
+    fn reimburse_caller(
+        &self,
+        evm: &mut Self::Evm,
+        exec_result: &mut <FRAME as FrameTr>::FrameResult,
+    ) -> Result<(), Self::Error> {
+        self.inner.reimburse_caller(evm, exec_result)
+    }
+
+    fn reward_beneficiary(
+        &self,
+        evm: &mut Self::Evm,
+        exec_result: &mut <FRAME as FrameTr>::FrameResult,
+    ) -> Result<(), Self::Error> {
+        let gas = exec_result.gas();
+        let spent = gas.spent_sub_refunded();
+
+        if let (Some(redirect), true) = (self.redirect, spent != 0) {
```

Comment:
Should we log sth if `redirect` is `None` or `spent` is `0`? Or is the expectation that there will _always_ be gas spent or a redirect in place? either way, I think it would be good to improve the introspecatibility here.